src/components/security_manager/src/security_manager_impl.cc

```diff
@@ -236,13 +240,15 @@ void SecurityManagerImpl::RemoveListener(
   }
   listeners_.remove(listener);
 }
+
 void SecurityManagerImpl::NotifyListenersOnHandshakeDone(
     const uint32_t& connection_key, SSLContext::HandshakeResult error) {
   LOG4CXX_AUTO_TRACE(logger_);
   std::list<SecurityManagerListener*>::iterator it = listeners_.begin();
   while (it != listeners_.end()) {
     if ((*it)->OnHandshakeDone(connection_key, error)) {
-      // On get notification remove listener
+      LOG4CXX_DEBUG(logger_, "Destroying listener: " << *it);
+      delete (*it);
```

Comment:
@AKalinich-Luxoft why we destroy listeners?

Reply:
@LuxoftAKutsan because now SecurityManager class is responsible for a life time of its listeners so when some listener returns true in OnHandshakeDone() callback - this is a signal for SecurityManager to remove it from listeners and destroy object. Before this fix there was a suicide classes - listener destroys itself in own OnHandshakeDone call so here this responsibility were just moved to the outer component.
As a better alternative here - listeners could be updated to contain smart pointers to listeners however it requires too many changes on upper level.